File: templates/problems/17_math_geometry/03_set_matrix_zeroes/set_matrix_zeroes_test.go
package set_matrix_zeroes

import (
	"reflect"
	"testing"
)

func TestSetZeroes(t *testing.T) {
	tests := []struct {
		name     string
		matrix   [][]int
		expected [][]int
	}{
		{"example 1", [][]int{{1, 1, 1}, {1, 0, 1}, {1, 1, 1}}, [][]int{{1, 0, 1}, {0, 0, 0}, {1, 0, 1}}},
		{"example 2", [][]int{{0, 1, 2, 0}, {3, 4, 5, 2}, {1, 3, 1, 5}}, [][]int{{0, 0, 0, 0}, {0, 4, 5, 0}, {0, 3, 1, 0}}},
		{"no zeros", [][]int{{1, 2}, {3, 4}}, [][]int{{1, 2}, {3, 4}}},
		{"all zeros", [][]int{{0, 0}, {0, 0}}, [][]int{{0, 0}, {0, 0}}},
		{"single element zero", [][]int{{0}}, [][]int{{0}}},
		{"single element nonzero", [][]int{{5}}, [][]int{{5}}},
		{"corner zero", [][]int{{0, 1}, {1, 1}}, [][]int{{0, 0}, {0, 1}}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			setZeroes(tt.matrix)
			if !reflect.DeepEqual(tt.matrix, tt.expected) {
				t.Errorf("setZeroes() = %v, want %v", tt.matrix, tt.expected)
			}
		})
	}
}

File: templates/problems/17_math_geometry/03_set_matrix_zeroes/typescript/setMatrixZeroes.test.ts
import { describe, it, expect } from 'vitest';
import { setZeroes } from './setMatrixZeroes';

describe('SetMatrixZeroes', () => {
    it('example 1', () => {
        const matrix = [[1, 1, 1], [1, 0, 1], [1, 1, 1]];
        setZeroes(matrix);
        expect(matrix).toEqual([[1, 0, 1], [0, 0, 0], [1, 0, 1]]);
    });

    it('example 2', () => {
        const matrix = [[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]];
        setZeroes(matrix);
        expect(matrix).toEqual([[0, 0, 0, 0], [0, 4, 5, 0], [0, 3, 1, 0]]);
    });

    it('no zeros', () => {
        const matrix = [[1, 2], [3, 4]];
        setZeroes(matrix);
        expect(matrix).toEqual([[1, 2], [3, 4]]);
    });

    it('all zeros', () => {
        const matrix = [[0, 0], [0, 0]];
        setZeroes(matrix);
        expect(matrix).toEqual([[0, 0], [0, 0]]);
    });

    it('single element zero', () => {
        const matrix = [[0]];
        setZeroes(matrix);
        expect(matrix).toEqual([[0]]);
    });

    it('single element nonzero', () => {
        const matrix = [[5]];
        setZeroes(matrix);
        expect(matrix).toEqual([[5]]);
    });

    it('corner zero', () => {
        const matrix = [[0, 1], [1, 1]];
        setZeroes(matrix);
        expect(matrix).toEqual([[0, 0], [0, 1]]);
    });
});
